listeners/listeners.go

```diff
@@ -25,115 +29,180 @@ type Transaction struct {
 }
 
-func (w *Listeners) AddBlockListener(id ids.ID, c BlockListener) {
-	w.blockL.Lock()
-	defer w.blockL.Unlock()
-
-	w.blockListeners[id] = c
-}
-
-func (w *Listeners) RemoveBlockListener(id ids.ID) {
-	w.blockL.Lock()
-	defer w.blockL.Unlock()
-
-	delete(w.blockListeners, id)
-}
-
 // If never possible for a tx to enter mempool, call this
-func (w *Listeners) RemoveTx(txID ids.ID, err error) {
+func (w *Listeners) RemoveTx(txID ids.ID, err error, streamingServer *pubsub.Server) {
 	w.txL.Lock()
 	defer w.txL.Unlock()
 
-	w.removeTx(txID, err)
+	w.removeTx(txID, err, streamingServer)
 }
 
-func (w *Listeners) removeTx(txID ids.ID, err error) {
+func (w *Listeners) removeTx(txID ids.ID, err error, streamingServer *pubsub.Server) {
 	listeners, ok := w.txListeners[txID]
 	if !ok {
 		return
 	}
-	for _, listener := range listeners {
-		select {
-		case listener <- &Transaction{
-			TxID: txID,
-			Err:  err,
-		}:
-		default:
-			// drop message if client is not keeping up or abandoned
-		}
-	}
+	p := codec.NewWriter(consts.MaxInt)
+	PackRemovedTxMessage(p, txID, err)
+	streamingServer.Publish([]byte(txID.String()), listeners)
 	delete(w.txListeners, txID)
 	// [expiringTxs] will be cleared eventually (does not support removal)
 }
 
-func (w *Listeners) SetMinTx(t int64) {
+func (w *Listeners) SetMinTx(t int64, streamingServer *pubsub.Server) {
 	w.txL.Lock()
 	defer w.txL.Unlock()
 
 	expired := w.expiringTxs.SetMin(t)
 	for _, id := range expired {
-		w.removeTx(id, ErrExpired)
+		w.removeTx(id, ErrExpired, streamingServer)
 	}
 }
 
-func (w *Listeners) AcceptBlock(b *chain.StatelessBlock) {
-	w.blockL.Lock()
-	for _, listener := range w.blockListeners {
-		select {
-		case listener <- b:
-		default:
-			// drop message if client is not keeping up or abandoned
-		}
-	}
-	w.blockL.Unlock()
+func (w *Listeners) AcceptBlock(
+	b *chain.StatelessBlock,
+	streamingServer *pubsub.Server,
+) {
+	p := codec.NewWriter(consts.MaxInt)
+	BlockMessageBytes(b, p)
+	// Publish accepted block to all block listeners
+	streamingServer.Publish(p.Bytes(), streamingServer.Conns())
 
 	w.txL.Lock()
 	defer w.txL.Unlock()
 	results := b.Results()
 	for i, tx := range b.Txs {
+		p := codec.NewWriter(consts.MaxInt)
 		txID := tx.ID()
 		listeners, ok := w.txListeners[txID]
 		if !ok {
 			continue
 		}
-		for _, listener := range listeners {
-			select {
-			case listener <- &Transaction{
-				TxID:   txID,
-				Result: results[i],
-			}:
-			default:
-				// drop message if client is not keeping up or abandoned
-			}
-		}
+		// Publish to tx listener
+		PackAcceptedTxMessage(p, txID, results[i])
+		streamingServer.Publish(
+			p.Bytes(),
+			listeners,
+		)
 		delete(w.txListeners, txID)
 		// [expiringTxs] will be cleared eventually (does not support removal)
 	}
 }
+
+// Could be a better place for these methods
+// Packs an accepted block message
+func PackAcceptedTxMessage(p *codec.Packer, txID ids.ID, result *chain.Result) {
```

Comment:
You should check for packer error at the end of these IMO.

Reply:
added for unpacking, checking during pack might be redundant since unpack checks for errors anyway?